### ISSUE TITLE
alsa: could not open auplay device 'default' (Device or resource busy)

### ISSUE THREAD

**Issue by dannymarosi** [baresip/baresip#1261]
Running BARESIP v1.0.0 on raspberry pi zero with buster distribution.
No audio when uncommenting “**ringback_disabled	no**” in config file.
![2021_03_10 RingBack issue](https://user-images.githubusercontent.com/62627034/110678053-b958b300-818a-11eb-9972-8ccffc51e911.png)
Thanks for your help.

**Comment by cspiel1** [comment]
I also use ALSA on pi zero. Will try to reproduce this.

But by checking our baresip branch I found that we have already patches that may solve this. With "normal" ALSA configuration the problem is, that the a device can only be opened by one instance. The ringback tone has to be stopped before the audio is opened for the established call.

When looking in master HEAD:
In call.c:1560 the audio is started for the call and then in line 1578 the CALL_EVENT_ESTABLISHED is passed to ua.c which forwards a UA-event to the menu. Then menu stops the ringback. In our branch we
- first stop the ringback, then
- start the audio for the call.

Another solution could be to configure ALSA differently. But I am not an expert. Maybe dmix and dsnoop plugins for ALSA help.

**Comment by alfredh** [comment]
there are also 4 180 ringing

is the audio player started 4 times?

**Comment by cspiel1** [comment]
With default configuration auplay_alloc() is called only once in play_tone().

The problem is that mem_deref(menu.play) is done too late. I will check our commits and create a PR if I am right.

**Comment by cspiel1** [comment]
@alfredh Maybe you saw my try to solve this in the mentioned PR. Should we live with this ALSA restriction?
Maybe add in module alsa doxygen a note?

On the other side, the current behavior also is not very good:
If the event handler terminates the call, we this scenario:
- Audio devices are opened (for the call).
- Call is terminated.
- Audio devices are closed.

**Comment by alfredh** [comment]
another option is to add a new event that is called before ESTABLISHED.

the menu module can then use this event to stop any ongoing tones

**Comment by cspiel1** [comment]
Right! And ESTABLISHED then could be sent by a timer with zero delay.

UA_EVENT_CALL_PROGRESS also stops the ringback tone. Outgoing call states are already more complicated than incoming.

OUTGOING --> RINGING or PROGRESS --> ESTABLISHED

PROGRESS (Early media) normally also starts audio (only the mic?) before the event is passed to upper layers. I am not sure but think this opens the Mic and should not concur with the ringback which opens playback device.


Another option: We could call call_stream_start() in a timer. In the destructor of the call we stop the timer.

**Comment by cspiel1** [comment]
I try a different PR: call_stream_start() is called asynchronously by timer.
The function start_audio() is called
- in call_stream_start() directly and
- in menc_event_handler() and stream_mnatconn_handler() both already asynchronusly.

**Comment by netyuchangfa** [comment]
I've had this problem before，then build and install alsalib ,alsa-lib-1.2.4.tar.bz2

**Comment by cspiel1** [comment]
@dannymarosi The fix is in master. Please cool you re-test?

**Comment by dannymarosi** [comment]
@cspiel1  I pulled the latest master commit (78083)
Then execute make && sudo make install
To test I had to remove the !(not) in line 437 on menu.c
ua_event_handler()

![2021_03_11 menu c](https://user-images.githubusercontent.com/62627034/110838289-25a0e880-8257-11eb-91a8-cd29f43992a9.png)

Unfortunately the problem still exist.
```
pi@raspberrypi:~/baresip $ baresip
baresip v1.0.0 Copyright (C) 2010 - 2020 Alfred E. Heggestad et al.
Local network address:  IPv4=wwan0|166.140.148.43
aucodec: PCMU/8000/1
aucodec: PCMA/8000/1
ausrc: alsa
auplay: alsa
medianat: stun
medianat: turn
medianat: ice
Populated 2 accounts
Populated 3 contacts
Populated 2 audio codecs
Populated 0 audio filters
Populated 0 video codecs
Populated 0 video filters
baresip is ready.
17778009107100@callcentric.com: (prio 0) {0/UDP/v4} 200 Ok () [1 binding]
dannymarosi@sip.linphone.org: (prio 0) {0/TCP/v4} 200 Registration successful (Flexisip/2.0.4/2.0)) [1 binding]
All 2 useragents registered successfully! (822 ms)
ua: using best effort AF: af=AF_INET
call: connecting to 'sip:19499887511@callcentric.com;transport=udp'..
call: SIP Progress: 100 Trying (/)
call: SIP Progress: 180 Ringing (/)
call: SIP Progress: 180 Ringing (/)
call: SIP Progress: 180 Ringing (/)
call: SIP Progress: 180 Ringing (/)
call: SIP Progress: 180 Ringing (/)
stream: update 'audio'
audio: Set audio decoder: PCMU 8000Hz 1ch
alsa: could not open auplay device 'default' (Device or resource busy)
audio: start_player failed (alsa.default): Unknown error -16
call: update: audio_decoder_set error: Unknown error -16
audio: Set audio encoder: PCMU 8000Hz 1ch
alsa: could not open auplay device 'default' (Device or resource busy)
audio: start_player failed (alsa.default): Unknown error -16
audio: source started with sample format S16LE
sip:19499887511@callcentric.com;transport=udp: session closed: Unknown error -16
Quit
ua: stop all (forced=0)
pi@raspberrypi:~/baresip $
```
Thanks for your help.

**Comment by dannymarosi** [comment]
Add clarification on change in menu.c
![2021_03_11 menu c2](https://user-images.githubusercontent.com/62627034/110842577-21c39500-825c-11eb-9fad-358346011ca6.png)

**Comment by cspiel1** [comment]
Please avoid uploading screen shots! Otherwise I have to do OCR to understand. ;-)

The use of has_outgoing_call() indeed is wrong. Thanks for the hint! But this is not the fix. Consider the case if the user starts multiple outgoing calls to different peers! Then the ringback tone should only be played for the first outgoing call. But the current call counts also to the outgoing calls. Its a mystery how this ever could work.

- I fix this right now.
- Then I try to reproduce the mentioned alsa failure. For that I have to disable the dmix plugin in the alsa config on my pi zero.

**Comment by cspiel1** [comment]
Fixed ringback tone and some other tone issues here:
https://github.com/baresip/baresip/pull/1273

**Comment by webstean** [comment]
Will this help the next person who get the "busy" message with alsa?
https://github.com/baresip/baresip/pull/1283

**Comment by cspiel1** [comment]
Yes, #1283 will help. We could take it. But should fix the issue anyway. In my opinion a default ALSA config should also work. And it should be possible. I will investigate it further.

**Comment by cspiel1** [comment]
Can reproduce the overlap of ringback and call audio. Most likely the reason for this issue.

The commit 2f45951f8841b99be2a5084796788febf55b61b9 does not help or it is not enough.

**Comment by cspiel1** [comment]
The fix for the sipsess_estab_handler() is needed.

But the call audio is also started from sipsess_answer_handler() -> update_media() -> update_audio() -> audio_decoder_set() and audio_encoder_set(). And the latter two both start the audio for the call. This is too early in my opinion. Will try to find a solution.

**Comment by webstean** [comment]
According to this: https://wiki.archlinux.org/index.php/Advanced_Linux_Sound_Architecture#Dmix 
dmix is the default configuration. Is this actually worth solving?

**Comment by cspiel1** [comment]
I am not sure if this is true for all platforms. There is another issue with dmix on raspbian: https://www.raspberrypi.org/forums/viewtopic.php?t=262071

@dannymarosi Which setup for audio_player do you use? You may test with this branch for the ringback: https://github.com/baresip/baresip/pull/1294. Or with your work-around. Please could you also try the setting sysdefault mentioned in the forum:
```
audio_player   alsa,sysdefault
audio_source  alsa,sysdefault
audio_alert     alsa,sysdefault
```

Do you have such alsa device on your PI zero?

**Comment by cspiel1** [comment]
I personally do not use raspbian. Instead ARM arch. And a special audio hat, but should be compatible to any other. My configuration is:
- baresip config
```
audio_player		alsa,ring
audio_source		alsa,micvol
audio_alert		alsa,ring
```
Attached my alsa config.
[asound.conf.zip](https://github.com/baresip/baresip/files/6148974/asound.conf.zip)

**Comment by cspiel1** [comment]
Reproducing on a dev host is difficult:
- ALSA may be configured to use pulseaudio
```
$ aplay -L
default
    Playback/recording through the PulseAudio sound server
...
```
- dmix may be used by default
- the soundcard may support HW mixing

**Comment by cspiel1** [comment]
I can reproduce the issue on my PI zero when setting baresip config:
```
audio_player		alsa,sysdefault
```

Above I posted the working configuration as stated already with a dmix configuration. Additionally, we have now the info
```
consider using dmix as your default alsa device
```

I belief a no dmix config is more common.

Solving the issue is difficult and error prune. As posted already:
> But the call audio is also started from sipsess_answer_handler() -> update_media() -> update_audio() -> audio_decoder_set() and audio_encoder_set(). And the latter two both start the audio for the call. This is too early in my opinion. 

A workaround is to use dmix. (Using dmix on a system without pulseaudio anyway has more benefits.) I am giving up here.
@alfredh Should we close this issue?

**Comment by juha-h** [comment]
Christian Spielberger writes:

> A workaround is to use dmix. (Using dmix on a system without
> pulseaudio anyway has more benefits.) I am giving up here.

Before pulseaudio, I always used dmix in my Linux alsa configuration.  I
don't see any need for fixes if alsa audio_source/player works with
dmix.

**Comment by dannymarosi** [comment]
@cspiel1 I use ALSA audio player on RASPBERRY OS 
Using dmix solved the issue for me.
See https://github.com/baresip/baresip/issues/832#issuecomment-547701825

Thanks for your help.

**Comment by cspiel1** [comment]
Thanks @dannymarosi for the feedback.

@juha-h You're right! But then I remembered that @alfredh already wrote a possible solution by adding a new call event. Here is a PR https://github.com/baresip/baresip/pull/1299 that does this and reverts the commit which did'nt solve this issue.
